### PR TITLE
[ONNX] Automatically convert dynamic_axes to dynamic_shapes with torch.export.Dim.AUTO

### LINKED ISSUE #143124: [ONNX] _rename_dynamic_shapes_with_model_inputs is not robust
### 🐛 Describe the bug

https://github.com/pytorch/pytorch/blob/520ba556cd4fcb80efc2849ca169bcfa8dfeff5e/torch/onnx/_internal/exporter/_compat.py#L58

The function was meant to align `dynamic_shapes` keys to model signature, since `torch.export.export` does not support customized `input_names`. However, model.signature could be the same length as input_names accidentally. For example, input_names includes all `kv_cache` names, and LLM models usually come with a lot of Optional inputs, while `kv_cache` names only counts one because it's a list.

### Versions

Nightly

### PR DESCRIPTION
**Pull request by titaiwangms**
With https://github.com/pytorch/pytorch/pull/133620 introducing Dim.AUTO, we can now automatically convert dynamic_axes to dynamic_shapes without specifying min and max. However, exporting still could be crashed when there are same specs shared between inputs and there is no guarantee that the axes will be dynamic (see PR description). 

~~Therefore, a~~ follow-up PR should create a post-processing ONNX side pass to ~~enable the missed dynamic axes~~ rename the dynamic shapes (s0,  s1, ...) to dynamic_axes (user setting names).

This PR does:
(1) Apply torch.export.Dim.AUTO to dynamic_axes when dynamic_shapes is not provided.
(2) Convert args/kwargs to tuple inputs, which follows the generated dynamic_shapes format to avoid errors during torch.export.export.
(3) Avoid KeyError in _rename_dynamic_shapes_with_model_inputs funtion.
(4) Add real world case of a HF model with kv_cache to test on ONNX exporter.

### CODE REVIEW

File: torch/onnx/_internal/exporter/_compat.py [hunk 1]
```diff
@@ -98,25 +101,18 @@ def _from_dynamic_axes_to_dynamic_shapes(
...
             dynamic_shapes[input_name] = {
-                k: torch.export.Dim(re.sub(r"[^A-Za-z_]", "", v), max=99999)
-                for k, v in axes.items()
+                k: torch.export.Dim.AUTO for k, _ in axes.items()
             }
         elif isinstance(axes, list):
-            dynamic_shapes[input_name] = {
-                k: torch.export.Dim(f"{input_name}_dim_{k}", max=99999) for k in axes
-            }
+            dynamic_shapes[input_name] = {k: torch.export.Dim.AUTO for k in axes}
```
Comment (gramalingam) [inline torch/onnx/_internal/exporter/_compat.py:115]: May be worth checking that `k` is a valid value (int in the expected range)?

File: torch/onnx/_internal/exporter/_compat.py [hunk 2]
```diff
@@ -98,25 +101,18 @@ def _from_dynamic_axes_to_dynamic_shapes(
...
             # User specified an output name as a dynamic axis, so we skip it
             continue
         if isinstance(axes, dict):
-            # Dim needs to pass str.isidentifier()
-            # If the max is not set, llm is going to fail, as sequence length is usually bounded within config.
-            # But we also don't want to only support llm. This kind of leaves us with this awkward position.
             dynamic_shapes[input_name] = {
-                k: torch.export.Dim(re.sub(r"[^A-Za-z_]", "", v), max=99999)
-                for k, v in axes.items()
+                k: torch.export.Dim.AUTO for k, _ in axes.items()
```
Comment (gramalingam) [inline torch/onnx/_internal/exporter/_compat.py:112]: nit: As below: may be worth checking k is a valid value.

File: test/onnx/exporter/test_small_models_e2e.py [hunk 3]
```diff
@@ -144,8 +147,243 @@ def forward(self, *x):
...
+        torch.nn.Module,
+        dict[str, Any],
+        dict[str, dict[int, str]],
+        list[str],
+        list[str],
+    ]
+):
+    model = GPTJForCausalLM.from_pretrained("hf-internal-testing/tiny-random-gptj")
+
+    input_ids = torch.tensor(
```
Comment (gramalingam) [inline test/onnx/exporter/test_small_models_e2e.py:180]: nit: are these specific values critical? are can we replace by random ints in a given range (which will take up less space, and clarify things better).

File: test/onnx/exporter/test_small_models_e2e.py [hunk 4]
```diff
@@ -144,8 +147,243 @@ def forward(self, *x):
...
+                523,
+                113,
+                288,
+                141,
+                516,
+                571,
+            ],
+        ]
+    )
+    attention_mask = torch.tensor(
```
Comment (gramalingam) [inline test/onnx/exporter/test_small_models_e2e.py:220]: nit: as above.

File: torch/onnx/_internal/exporter/_compat.py [hunk 5]
```diff
@@ -98,25 +101,18 @@ def _from_dynamic_axes_to_dynamic_shapes(
 
     dynamic_shapes: dict[str, Any | None] = {}
     for input_name, axes in dynamic_axes.items():
-        # NOTE: torch.export.Dim requires strict min and max constraints, and it
-        # dpends on the traced model to provide the correct min and max values.
-        # We set max to 99999 to avoid the constraints violation error with the default int64 max.
-        # https://github.com/pytorch/pytorch/blob/32f585d9346e316e554c8d9bf7548af9f62141fc/test/export/test_export.py#L687
+        # TODO(titaiwang): Add ONNX IR pass to post-process the dynamic axes
```
Comment (gramalingam) [inline torch/onnx/_internal/exporter/_compat.py:104]: Is it possible to include the information returned by AUTO dim analysis in the export-report? For example, if analysis indicates dim1 and dim2 are expected to be equal? Or, if it indicates dim1 is expected to be in range (5, 10)?

File: test/onnx/exporter/test_small_models_e2e.py (hunk 3 above)
Comment (titaiwangms) [reply test/onnx/exporter/test_small_models_e2e.py]: I was concerned about flaky, but we will see.

File: test/onnx/exporter/test_small_models_e2e.py [hunk 6]
```diff
@@ -144,8 +147,143 @@ def forward(self, *x):
         onnx_program = torch.onnx.export(VisionModel(), args, dynamo=True)
         onnx_testing.assert_onnx_program(onnx_program)
 
+    def test_onnx_export_huggingface_llm_models_with_kv_cache(self):
```
Comment (justinchuby) [inline test/onnx/exporter/test_small_models_e2e.py:150]: This file is meant to house end to end tests for "small" models whose definition being just a few ops. A HF model should be placed elsewhere. Consider creating a file just for testing HF models to make change management easier.

File: test/onnx/exporter/test_small_models_e2e.py [hunk 7]
```diff
@@ -3,7 +3,10 @@
 
 from __future__ import annotations
 
+from typing import Any
+
 import torchvision
+from transformers import GPTJForCausalLM
```
Comment (justinchuby) [inline test/onnx/exporter/test_small_models_e2e.py:9]: Import modules only. Consider moving to a different test file as commented below.

File: torch/onnx/_internal/exporter/_compat.py (hunk 1 above)
Comment (titaiwangms) [reply torch/onnx/_internal/exporter/_compat.py:124]: Done

File: torch/onnx/_internal/exporter/_compat.py (hunk 5 above)
Comment (titaiwangms) [reply torch/onnx/_internal/exporter/_compat.py]: This can a follow-up PR that we try to identify the constraints and store them in metadata of ONNX IR.

File: test/onnx/exporter/test_small_models_e2e.py (hunk 6 above)
Comment (titaiwangms) [reply test/onnx/exporter/test_small_models_e2e.py]: Done

Comment (titaiwangms) [reply test/onnx/exporter/test_small_models_e2e.py]: Created test_hf_models_e2e.py for llm testing.

File: test/onnx/exporter/test_hf_models_e2e.py [hunk 8]
```diff
@@ -0,0 +1,155 @@
...
+def _prepare_llm_model_gptj_to_test() -> (
+    tuple[
+        torch.nn.Module,
+        dict[str, Any],
+        dict[str, dict[int, str]],
+        list[str],
+        list[str],
+    ]
+):
+    model = transformers.GPTJForCausalLM.from_pretrained(
```
Comment (xadupre) [inline test/onnx/exporter/test_hf_models_e2e.py:42]: Do you know the size of it?

**Review by xadupre — APPROVED** [review]

**Comment by titaiwangms** [comment]
@xadupre Could you approve again to see if you have the ownership now?

**Review by xadupre — APPROVED** [review]

**Review by shubhambhokare1 — APPROVED** [review]
